File: generators/ee3_generator.py
from datetime import datetime
from io import BytesIO
from PyPDF2 import PdfReader, PdfWriter
from reportlab.pdfgen import canvas
from reportlab.lib.pagesizes import letter


class EE3Generator:
    def __init__(self, template_path="templates/EE-3.pdf"):
        self.template_path = template_path

    def format_date(self, date_value, format_str="%m/%d/%Y"):
        """Helper function to format dates that could be strings or datetime objects"""
        if not date_value:
            return ""

        if isinstance(date_value, str):
            try:
                date_obj = datetime.strptime(date_value, "%Y-%m-%d")
                return date_obj.strftime(format_str)
            except ValueError:
                try:
                    date_obj = datetime.strptime(date_value, "%m/%d/%Y")
                    return date_obj.strftime(format_str)
                except ValueError:
                    return str(date_value)
        else:
            return date_value.strftime(format_str)

    def wrap_text(self, text, max_chars_per_line=140, max_lines=4):
        """Wrap text into multiple lines with character limit"""
        if not text:
            return [""]

        words = text.split(" ")
        lines = []
        current_line = ""

        for word in words:
            if len(current_line + word) <= max_chars_per_line:
                current_line += word + " "
            else:
                if current_line:
                    lines.append(current_line.strip())
                current_line = word + " "

                if len(lines) >= max_lines:
                    print(f"[EE3] Warning: Text exceeds {max_lines} lines. Text truncated.")
                    return lines

        if current_line:
            lines.append(current_line.strip())

        if len(lines) > max_lines:
            return lines[:max_lines]

        return lines

    def generate(self, client_record: dict, form_data: dict) -> tuple[str, BytesIO]:
        first_name = form_data.get("first_name", "")
        last_name = form_data.get("last_name", "")
        former_name = form_data.get("former_name", "")
        ssn = form_data.get("ssn", "")
        employment_history = form_data.get("employment_history", [])

        # Limit to 3 employers
        if len(employment_history) > 3:
            print("[EE3] Warning: EE-3 form supports maximum 3 employers. Only first 3 will be included.")
            employment_history = employment_history[:3]

        # Generate filename
        current_date = datetime.now().strftime("%m.%d.%y")
        if first_name and last_name:
            filename = f"EE3_{first_name[0]}.{last_name}_{current_date}.pdf"
        else:
            filename = f"EE3_Unknown_{current_date}.pdf"

        # Generate PDF
        pdf_bytes = self.draw_pdf(
            first_name,
            last_name,
            former_name,
            ssn,
            employment_history,
        )
        return filename, pdf_bytes

    def draw_employer_section(self, overlay, job, base_y, deltas, page_num):
        """Draw a single employer section on the overlay"""
        # Parse dates
        start_date = job.get("start_date")
        end_date = job.get("end_date")

        start_date_str = self.format_date(start_date) if start_date else ""
        end_date_str = self.format_date(end_date) if end_date else ""

        # Dates section
        dates_y = base_y + deltas["dates"]
        if start_date_str:
            parts = start_date_str.split("/")
            if len(parts) == 3:
                overlay.drawString(175, dates_y, parts[0])  # month
                overlay.drawString(205, dates_y, parts[1])  # day
                overlay.drawString(230, dates_y, parts[2])  # year

        if end_date_str:
            parts = end_date_str.split("/")
            if len(parts) == 3:
                overlay.drawString(360, dates_y, parts[0])  # month
                overlay.drawString(390, dates_y, parts[1])  # day
                overlay.drawString(415, dates_y, parts[2])  # year

        # Facility information
        facility_y = base_y + deltas["facility"]
        overlay.drawString(30, facility_y, job.get("facility_name", ""))
        overlay.drawString(265, facility_y, job.get("specific_location", ""))

        # City/State
        city = job.get("city", "")
        state = job.get("state", "")
        city_state = f"{city}, {state}" if city and state else city or state
        overlay.drawString(435, facility_y, city_state)

        # Contractor
        contractor_y = base_y + deltas["contractor"]
        overlay.drawString(30, contractor_y, job.get("contractor", ""))

        # Position Title
        position_y = base_y + deltas["position"]
        overlay.drawString(30, position_y, job.get("position_title", ""))

        # Union Member checkbox
        if job.get("union_member") and "union_checkbox" in deltas:
            union_y = base_y + deltas["union_checkbox"]
            overlay.drawString(204, union_y, "X")

        # Dosimetry Badge Worn checkbox
        if job.get("dosimetry_worn") and "dosimetry_checkbox" in deltas:
            dosimetry_y = base_y + deltas["dosimetry_checkbox"]
            overlay.drawString(442, dosimetry_y, "X")

        # Facility Type Checkbox - DOE Facility (always mark)
        if "facility_checkbox" in deltas:
            facility_checkbox_y = base_y + deltas["facility_checkbox"]
            overlay.drawString(238, facility_checkbox_y, "X")

        # Work Duties
        duties_y = base_y + deltas["duties"]
        duties_text = job.get("work_duties", "")
        duties_lines = self.wrap_text(duties_text)

        if duties_lines and duties_lines[0]:
            y_position = duties_y
            for line in duties_lines:
                overlay.drawString(20, y_position, line)
                y_position -= 12

        # Exposures (only for page 2)
        if "exposures" in deltas:
            exposures_y = base_y + deltas["exposures"]
            exposures_text = "Claimant stated they were exposed to radiation, silica dust, and other chemicals, solvents and contaminants during the course of their employment."
            exposures_lines = self.wrap_text(exposures_text)

            if exposures_lines and exposures_lines[0]:
                y_position = exposures_y
                for line in exposures_lines:
                    overlay.drawString(20, y_position, line)
                    y_position -= 12

    def draw_pdf(
        self,
        first_name: str,
        last_name: str,
        former_name: str,
        ssn: str,
        employment_history: list,
    ) -> BytesIO:
        # Load the existing PDF
        base_pdf = PdfReader(self.template_path)
        output = PdfWriter()

        current_date = datetime.now().strftime("%m/%d/%Y")

        # Define page-specific coordinate deltas
        page1_deltas = {
            "dates": 0,
            "facility": -42,
            "contractor": -87,
            "position": -119,
            "facility_checkbox": -70,
            "union_checkbox": -420,
            "dosimetry_checkbox": -114,
            "duties": -180,
            "exposures": -286,
        }

        page2_employer2_deltas = {
            "dates": 0,
            "facility": -38,
            "contractor": -79,
            "position": -108,
            "duties": -160,
            "exposures": -212,
            "facility_checkbox": -65,
            "union_checkbox": -282,
            "dosimetry_checkbox": -107,
        }

        page2_employer3_deltas = {
            "dates": 0,
            "facility": -37,
            "contractor": -80,
            "position": -110,
            "duties": -163,
            "exposures": -213,
            "facility_checkbox": -66,
            "union_checkbox": -286,
            "dosimetry_checkbox": -109,
        }

        # Employer positions
        employer_positions = [
            {"page": 0, "base_y": 468, "deltas": page1_deltas},
            {"page": 1, "base_y": 761, "deltas": page2_employer2_deltas},
            {"page": 1, "base_y": 460, "deltas": page2_employer3_deltas},
        ]

        # --- Page 1 ---
        page1_overlay_buffer = BytesIO()
        page1_overlay = canvas.Canvas(page1_overlay_buffer, pagesize=letter)
        page1_overlay.setFont("Helvetica", 9)

        # Header info - Employee Name (Field 1)
        page1_overlay.drawString(30, 640, last_name)
        page1_overlay.drawString(150, 640, first_name)
        if former_name:
            page1_overlay.drawString(250, 640, former_name)
        page1_overlay.drawString(440, 640, ssn)

        # Draw employer 1 if exists
        if len(employment_history) > 0:
            pos = employer_positions[0]
            self.draw_employer_section(
                page1_overlay,
                employment_history[0],
                pos["base_y"],
                pos["deltas"],
                0
            )

        page1_overlay.save()
        page1_overlay_buffer.seek(0)

        # Merge page 1
        page1_overlay_pdf = PdfReader(page1_overlay_buffer)
        base_page1 = base_pdf.pages[0]
        base_page1.merge_page(page1_overlay_pdf.pages[0])
        output.add_page(base_page1)

        # --- Page 2 ---
        if len(base_pdf.pages) > 1:
            page2_overlay_buffer = BytesIO()
            page2_overlay = canvas.Canvas(page2_overlay_buffer, pagesize=letter)
            page2_overlay.setFont("Helvetica", 9)

            # Date on page 2
            page2_overlay.drawString(385, 60, current_date)

            # Draw employers 2 and 3 if they exist
            for i in range(1, min(len(employment_history), 3)):
                pos = employer_positions[i]
                self.draw_employer_section(
                    page2_overlay,
                    employment_history[i],
                    pos["base_y"],
                    pos["deltas"],
                    1
                )

            page2_overlay.save()
            page2_overlay_buffer.seek(0)

            page2_overlay_pdf = PdfReader(page2_overlay_buffer)
            base_page2 = base_pdf.pages[1]
            base_page2.merge_page(page2_overlay_pdf.pages[0])
            output.add_page(base_page2)

        # --- Page 3 ---
        if len(base_pdf.pages) > 2:
            page3_overlay_buffer = BytesIO()
            page3_overlay = canvas.Canvas(page3_overlay_buffer, pagesize=letter)
            page3_overlay.setFont("Helvetica", 9)
            page3_overlay.drawString(500, 50, current_date)
            page3_overlay.save()
            page3_overlay_buffer.seek(0)

            page3_overlay_pdf = PdfReader(page3_overlay_buffer)
            base_page3 = base_pdf.pages[2]
            base_page3.merge_page(page3_overlay_pdf.pages[0])
            output.add_page(base_page3)

        result = BytesIO()
        output.write(result)
        result.seek(0)
        return result

File: web/src/lib/generators/ee3-generator.ts
/**
 * EE-3 Form Generator
 * Converts Python ee3_generator.py to TypeScript
 * MOST COMPLEX: Multi-page with text wrapping, page-specific coordinate deltas, 3 employer sections
 */

import { PDFDocument, PDFPage, StandardFonts } from "pdf-lib";
import { BaseGenerator } from "./base-generator";
import { ClientRecord, GeneratorResult } from "./types";
import { formatDateMMDDYY, formatDateMMDDYYYY } from "./utils/formatters";

export interface EE3EmploymentRecord {
  start_date: string; // ISO date
  end_date: string; // ISO date
  facility_name: string;
  specific_location: string;
  city: string;
  state: string;
  contractor: string;
  position_title: string;
  union_member?: boolean;
  dosimetry_worn?: boolean;
  work_duties: string;
}

export interface EE3ContactPerson {
  first_name: string;
  last_name: string;
  address: string;
  city: string;
  state: string;
  zip: string;
}

export interface EE3EmployeeContact {
  address: string;
  city: string;
  state: string;
  zip: string;
  phone_home: string;
  phone_work: string;
  phone_cell: string;
}

export interface EE3FormData {
  first_name: string;
  last_name: string;
  former_name?: string;
  ssn: string;
  employment_history: EE3EmploymentRecord[];
  employee_address: string;
  employee_city: string;
  employee_state: string;
  employee_zip: string;
  phone_home: string;
  phone_work: string;
  phone_cell: string;
  contact_first_name: string;
  contact_last_name: string;
  contact_address: string;
  contact_city: string;
  contact_state: string;
  contact_zip: string;
}

interface CoordinateDeltas {
  dates: number;
  facility: number;
  contractor: number;
  position: number;
  union_checkbox?: number;
  dosimetry_checkbox?: number;
  facility_checkbox?: number;
  duties: number;
  exposures?: number;
}

export class EE3Generator extends BaseGenerator {
  constructor() {
    super("EE-3.pdf");
  }

  /**
   * Wrap text into multiple lines with character limit
   * Returns empty array if text exceeds max_lines (warning logged)
   */
  private wrapText(
    text: string,
    maxCharsPerLine: number = 140,
    maxLines: number = 4
  ): string[] {
    if (!text) return [""];

    const words = text.split(" ");
    const lines: string[] = [];
    let currentLine = "";

    for (const word of words) {
      // Check if adding this word would exceed the character limit
      if ((currentLine + word).length <= maxCharsPerLine) {
        currentLine += word + " ";
      } else {
        // Finish current line and start new one
        if (currentLine) {
          lines.push(currentLine.trim());
        }
        currentLine = word + " ";

        // Check if we've exceeded max lines
        if (lines.length >= maxLines) {
          console.warn(
            `[EE3] Warning: Work duties text exceeds ${maxLines} lines. Text not added to PDF - please add manually after generation.`
          );
          return [];
        }
      }
    }

    // Add the last line
    if (currentLine) {
      lines.push(currentLine.trim());
    }

    // Final check for line count
    if (lines.length > maxLines) {
      console.warn(
        `[EE3] Warning: Work duties text exceeds ${maxLines} lines. Text not added to PDF - please add manually after generation.`
      );
      return [];
    }

    return lines;
  }

  /**
   * Draw a single employer section directly on a page
   */
  private drawEmployerSectionDirect(
    page: PDFPage,
    job: EE3EmploymentRecord,
    baseY: number,
    deltas: CoordinateDeltas,
    totalEmployers: number,
    pageNum: number
  ): void {
    // Parse dates
    let startDateStr = "";
    if (job.start_date) {
      const startDate = new Date(job.start_date);
      const month = String(startDate.getMonth() + 1).padStart(2, "0");
      const day = String(startDate.getDate()).padStart(2, "0");
      const year = String(startDate.getFullYear());
      startDateStr = `${month}/${day}/${year}`;
    }

    let endDateStr = "";
    if (job.end_date) {
      const endDate = new Date(job.end_date);
      const month = String(endDate.getMonth() + 1).padStart(2, "0");
      const day = String(endDate.getDate()).padStart(2, "0");
      const year = String(endDate.getFullYear());
      endDateStr = `${month}/${day}/${year}`;
    }

    // Dates section
    const datesY = baseY + deltas.dates;
    if (startDateStr) {
      const [month, day, year] = startDateStr.split("/");
      page.drawText(month, { x: 175, y: datesY, size: 9 });
      page.drawText(day, { x: 205, y: datesY, size: 9 });
      page.drawText(year, { x: 230, y: datesY, size: 9 });
    }

    if (endDateStr) {
      const [month, day, year] = endDateStr.split("/");
      page.drawText(month, { x: 360, y: datesY, size: 9 });
      page.drawText(day, { x: 390, y: datesY, size: 9 });
      page.drawText(year, { x: 415, y: datesY, size: 9 });
    }

    // Facility information
    const facilityY = baseY + deltas.facility;
    page.drawText(job.facility_name, { x: 30, y: facilityY, size: 9 });
    page.drawText(job.specific_location, { x: 265, y: facilityY, size: 9 });

    // City/State
    const cityState = `${job.city}, ${job.state}`;
    page.drawText(cityState, { x: 435, y: facilityY, size: 9 });

    // Contractor/sub-contractor
    const contractorY = baseY + deltas.contractor;
    page.drawText(job.contractor, { x: 30, y: contractorY, size: 9 });

    // Position Title
    const positionY = baseY + deltas.position;
    page.drawText(job.position_title, { x: 30, y: positionY, size: 9 });

    // Union Member checkbox
    const unionMember = job.union_member || false;
    if (deltas.union_checkbox !== undefined && unionMember) {
      const unionCheckboxY = baseY + deltas.union_checkbox;
      page.drawText("X", { x: 204, y: unionCheckboxY, size: 10 });
    }

    // Dosimeter Badge Worn checkbox
    const dosimetryWorn = job.dosimetry_worn || false;
    if (deltas.dosimetry_checkbox !== undefined && dosimetryWorn) {
      const dosimetryCheckboxY = baseY + deltas.dosimetry_checkbox;
      page.drawText("X", { x: 442, y: dosimetryCheckboxY, size: 10 });
    }

    // Facility Type Checkboxes - Department of Energy Facility
    if (deltas.facility_checkbox !== undefined) {
      const facilityCheckboxY = baseY + deltas.facility_checkbox;
      // Department of Energy Facility checkbox - always mark
      page.drawText("X", { x: 238, y: facilityCheckboxY, size: 10 });
    }

    // Description of Work Duties
    const dutiesY = baseY + deltas.duties;
    const duties = job.work_duties;
    const dutiesLines = this.wrapText(duties);

    // Only add work duties if they fit within the line limit
    if (dutiesLines.length > 0) {
      let yPosition = dutiesY;
      for (const line of dutiesLines) {
        page.drawText(line, { x: 20, y: yPosition, size: 9 });
        yPosition -= 12; // Move down 12 points for next line
      }
    }

    // Work conditions/exposures - only for page 2
    if (deltas.exposures !== undefined) {
      const exposuresY = baseY + deltas.exposures;
      const exposuresText =
        "Claimant stated they were exposed to radiation, silica dust, and other chemicals, solvents and contaminants during the course of their employment.";
      const exposuresLines = this.wrapText(exposuresText);

      if (exposuresLines.length > 0) {
        let yPosition = exposuresY;
        for (const line of exposuresLines) {
          page.drawText(line, { x: 20, y: yPosition, size: 9 });
          yPosition -= 12;
        }
      }
    }
  }


  async generate(
    clientRecord: ClientRecord,
    doctor: string,
    formData: EE3FormData
  ): Promise<GeneratorResult> {
    const firstName = formData.first_name || "";
    const lastName = formData.last_name || "";
    const name = lastName && firstName ? `${lastName}, ${firstName}` : "";
    const formerName = formData.former_name || "";
    const ssn = formData.ssn || "";
    let employmentHistory = formData.employment_history || [];

    // Check if we have more employers than supported
    if (employmentHistory.length > 3) {
      console.warn(
        `[EE3] Warning: EE-3 form supports maximum 3 employers. Only first 3 will be included. Additional employers need to be added manually.`
      );
      employmentHistory = employmentHistory.slice(0, 3);
    }

    // Generate filename
    const currentDate = formatDateMMDDYY();
    const filename =
      firstName && lastName
        ? `EE3_${firstName[0]}.${lastName}_${currentDate}.pdf`
        : `EE3_Unknown_${currentDate}.pdf`;

    // Load base template
    const basePdfDoc = await this.loadTemplate();

    // Define page-specific coordinate deltas
    const page1Deltas: CoordinateDeltas = {
      dates: 0,
      facility: -42,
      contractor: -87,
      position: -119,
      facility_checkbox: -70,
      union_checkbox: -420,
      dosimetry_checkbox: -114,
      duties: -180,
      exposures: -286,
    };

    const page2Employer2Deltas: CoordinateDeltas = {
      dates: 0,
      facility: -38,
      contractor: -79,
      position: -108,
      duties: -160,
      exposures: -212,
      facility_checkbox: -65,
      union_checkbox: -282,
      dosimetry_checkbox: -107,
    };

    const page2Employer3Deltas: CoordinateDeltas = {
      dates: 0,
      facility: -37,
      contractor: -80,
      position: -110,
      duties: -163,
      exposures: -213,
      facility_checkbox: -66,
      union_checkbox: -286,
      dosimetry_checkbox: -109,
    };

    // Define employer positions
    interface EmployerPosition {
      page: number;
      baseY: number;
      deltas: CoordinateDeltas;
    }

    const employerPositions: EmployerPosition[] = [
      { page: 0, baseY: 468, deltas: page1Deltas }, // Employer 1 - Page 1
      { page: 1, baseY: 761, deltas: page2Employer2Deltas }, // Employer 2 - Page 2 top
      { page: 1, baseY: 460, deltas: page2Employer3Deltas }, // Employer 3 - Page 2 bottom
    ];

    const formattedCurrentDate = formatDateMMDDYYYY();

    // Instead of creating separate overlay documents and embedding them,
    // we'll copy the template pages and draw directly on them
    const finalDoc = await PDFDocument.create();

    // Copy all pages from base template
    const copiedPages = await finalDoc.copyPages(
      basePdfDoc,
      basePdfDoc.getPageIndices()
    );

    // Page 1
    const page1 = copiedPages[0];
    const font = await finalDoc.embedFont(StandardFonts.Helvetica);
    page1.setFont(font);

    // Copy all the drawing operations from page1Overlay to page1
    page1.drawText(name, { x: 30, y: 640, size: 9 });
    if (formerName) {
      page1.drawText(formerName, { x: 250, y: 640, size: 9 });
    }
    page1.drawText(ssn, { x: 440, y: 640, size: 9 });

    // Draw employer 1 if exists
    if (employmentHistory.length > 0) {
      const position = employerPositions[0];
      this.drawEmployerSectionDirect(
        page1,
        employmentHistory[0],
        position.baseY,
        position.deltas,
        employmentHistory.length,
        position.page
      );
    }

    finalDoc.addPage(page1);

    // Page 2
    let page2: PDFPage;
    if (basePdfDoc.getPageCount() > 1) {
      page2 = copiedPages[1];
    } else {
      page2 = finalDoc.addPage([612, 792]);
    }

    page2.setFont(font);
    page2.drawText(formattedCurrentDate, { x: 385, y: 60, size: 9 });

    // Draw employers 2 and 3 if they exist
    for (let i = 1; i < employmentHistory.length && i < 3; i++) {
      const position = employerPositions[i];
      this.drawEmployerSectionDirect(
        page2,
        employmentHistory[i],
        position.baseY,
        position.deltas,
        employmentHistory.length,
        position.page
      );
    }

    finalDoc.addPage(page2);

    // Page 3 if exists
    if (basePdfDoc.getPageCount() > 2) {
      const page3 = copiedPages[2];
      page3.setFont(font);
      page3.drawText(formattedCurrentDate, { x: 500, y: 50, size: 9 });
      finalDoc.addPage(page3);
    }

    // Save PDF
    const pdfBytes = await finalDoc.save();

    return {
      filename,
      pdfBytes: Buffer.from(pdfBytes),
    };
  }
}
